feat: Email Verification upon Signup

solderq35 opened the pull request:
When joining new team

solderq35 commented:
Demo Vid: https://youtube.com/shorts/e6q_DAK7Q_g

solderq35 commented:
TODO: 
- [ ] Find a better way to update currentUser.emailVerified
- [ ] Use Alert.js file for alert or dialogs, don't use console.error

solderq35 commented:
I think the current changes in Auth.js are extraneous and can be removed

FrankreedX commented:
> Find a better way to update currentUser.emailVerified

Maybe onAuthStateChanged?

solderq35 commented:
New Demo Vid: https://youtu.be/yzMtmMPuigU

Intended Workflow: New User signs up (email verification sent) > User taps "Join Team" button (or taps "Resend Verification Email" if needed) > User is logged into the app and redirected to homescreen (`onIdTokenChanged` asynchronously watches for when the email is verified on the user server side)

solderq35 commented:
PS does the "join team" and "resend email verification" buttons need debouncing? I noticed that "join team" seemed to trigger multiple times sometimes (like 3 times in a row rapidly, hence janky navigation after tapping "join team" button).

FrankreedX commented:
> PS does the "join team" and "resend email verification" buttons need debouncing? I noticed that "join team" seemed to trigger multiple times sometimes (like 3 times in a row rapidly, hence janky navigation after tapping "join team" button).

Yes, and it will be addressed in either debounce 2 (if simple) or debounce-n (if complex) once this PR is merged

FrankreedX commented on app/(auth)/signup.js:
```diff
@@ -62,6 +66,17 @@ export default function SignUp() {
       });
 
+      sendEmailVerification(auth.currentUser)
+        .then(() => {
```
I think that we should be consistent with other code by using await here

solderq35 replied:
Should be resolved (moved to try / await syntax)

FrankreedX commented on app/segments/(team)/chooseTeam.js:
```diff
@@ -99,22 +104,56 @@ function ChooseTeam() {
-                router.replace("/");
+                let completed = false;
+
+                const unregisterAuthObserver = onIdTokenChanged(
```
Nice job finding out a different observer for this! Shouldn't this observer be registered when this page is loaded?

solderq35 replied:
Is the best way to do that to put the code in a useEffect or something, like in `context/Auth.js`?
- https://github.com/Golf-Drill-Challenge-App/Golf-App/blob/main/context/Auth.js#L41

I set up the observer to only run when the user pressed the button as I was paranoid about it executing too often etc.

FrankreedX replied:
You can pass it an empty array so it only executes once. Although that has failed me once with the debounce timer thingie...

FrankreedX commented on app/segments/(team)/chooseTeam.js:
```diff
@@ -99,22 +104,56 @@ function ChooseTeam() {
-                  ["userInfo", { userId: currentUserId }],
-                ]);
-                router.replace("/");
+                let completed = false;
```
Is this completed used to debounce the "Join team" button in case of a success?

solderq35 replied:
I was trying to prevent navigation redirect to home ("/") from triggering multiple times. I think it worked but it's pretty jank and a general debouncing PR on the button itself should also do the job.

FrankreedX commented on app/segments/(team)/chooseTeam.js:
```diff
@@ -99,22 +104,56 @@ function ChooseTeam() {
+                );
+
+                // Trigger a user reload to update the token and invoke the listener
+                await auth.currentUser.reload();
```
I like your placement of the reload here. Should we also place one in a pull to refresh for this page?

solderq35 replied:
Yes, I think it would be a good idea to add pulldown refresh

FrankreedX commented on app/segments/(team)/chooseTeam.js:
```diff
@@ -99,22 +104,56 @@ function ChooseTeam() {
+                );
+
+                // Trigger a user reload to update the token and invoke the listener
+                await auth.currentUser.reload();
```
I like your placement of the reload here. Should we also place one in a pull to refresh for this page?

solderq35 replied:
duplicate comment?

FrankreedX commented on app/segments/(team)/chooseTeam.js:
```diff
@@ -141,6 +180,35 @@ function ChooseTeam() {
             justifyContent: "center",
           }}
         >
+          <Button
```
Maybe we can not display the "Join team" button until the user is verified?

solderq35 replied:
Sure, I don't mind hiding that button

FrankreedX commented on app/segments/(team)/chooseTeam.js:
```diff
@@ -99,22 +104,56 @@ function ChooseTeam() {
+                    if (user && !completed) {
+                      await user.reload();
+                      if (user.emailVerified) {
+                        completed = true;
```
IIRC the onAuthStateChanged in Auth.js only fires when you login or signup and not on token change. Do we need to set auth.currentUserVerified to true here? Maybe we'll need this variable in the code in the future?

solderq35 replied:
I didn't realize that context/Auth.js also has a an emailVerified property. Yeah, it should be better to combine them somehow (not sure).

FrankreedX replied:
from my experience with displayName, signup.js sets the displayname on firebase but not locally, so I had to also set it locally. If you can use your onTokenChange observer to handle this it'll be great

FrankreedX reviewed (changes requested):
Works. I encountered an issue where verification success is printed twice.
![Screenshot 2024-07-23 at 2 44 38 PM](https://github.com/user-attachments/assets/9c492d48-4398-4171-96fd-14ee5d5cef21)

FrankreedX commented:
What is your plan to add button spinners for all of this PR's operations?

solderq35 commented:
> What is your plan to add button spinners for all of this PR's operations?

Assuming that "Join Team" button is hidden to user until email is verified, I guess we could make the "Resend Verification Email" button show a spinner animation until the email is finished sending?

For verified users, I suppose the "Join Team" button could show a spinner animation until all async operations were awaited (firestore operations etc)

FrankreedX commented:
Ye it took a very long time for me when I tested it. It's helpful to know that the app received my input and is doing work instead of just trusting it.

solderq35 commented:
### 07-25-2024 Progress
- [ ] Pulldown refresh
- [ ] Hide "join team" button until user is verified?
- [ ] spinner animation ("Loading") while Verification email is being sent

### Screenshot
![image](https://github.com/user-attachments/assets/86f2a16d-b9fe-4c2c-814e-29889c7cd782)

solderq35 commented:
Fixed the below signout error (see screenshot), although not sure if this will cause excess re renders etc, maybe there is better solution.

![image](https://github.com/user-attachments/assets/c2d61f2a-450f-480a-8b7d-706083166f6d)

solderq35 commented:
![image](https://github.com/user-attachments/assets/fea32fc0-7ffc-4677-8f54-f5222f2d7ec8)

Kind of got loading spinner working for sending verification email, and also kind of fixed bug where it was inconsistent if the new user would be directed to chooseTeam screen upon signup.

TODO:
- Cleanup code / log statements
- Maybe switch to `auth.currentUser.emailVerified` as it seems to work more reliably (given persistent login credentials)

FrankreedX commented:
you gotta set the "textColor" for the spinner to change

solderq35 commented:
Changed spinner color to white:
![image](https://github.com/user-attachments/assets/da82651c-644c-4c34-9f44-15683f630033)

I removed the dependency array from the `unregisterAuthObserver` useEffect due to it bugging out.

Basically you need to call `user.reload()` in order for the updated status to register. If you also put `auth.currentUser` (or whatever) in the useEffect dependency array, it causes an infinite feedback loop.

So now it does work but the user has to manually pulldown to refresh after they verified email

Maybe helpful for the useEffect / emailVerified / onIdTokenChanged issue:
- https://stackoverflow.com/questions/76961591/firebase-is-there-any-observer-to-watch-for-change-in-email-verification-status
- https://github.com/firebase/firebase-js-sdk/issues/3817
- https://stackoverflow.com/questions/53508364/update-the-email-verification-status-without-reloading-page

solderq35 commented:
I think the useEffect works as intended now (it automatically detects when you verified email and then shows "Join Team" button afterwards).

To resolve the issues mentioned in my previous comment, I had to set up a setInterval to reload the user and check for updates every 10 seconds.

Demo Video (read Youtube video description): https://youtu.be/KpHN71xhFfE

solderq35 commented:
PS: I tried putting `showSnackBar` in the useEffect dependency Array, but it caused the "Email successfully verified" popup to not go away for some reason later, especially if I had waited about 20 or 30 seconds after signup to verify email.

So I am leaving it out of the dependency array for now.

![image](https://github.com/user-attachments/assets/5321874a-934e-4dee-a0fb-312b06a93e07)

Gehrkej commented on app/(auth)/signup.js:
```diff
@@ -66,6 +70,15 @@ export default function SignUp() {
       });
 
+      try {
+        await sendEmailVerification(auth.currentUser);
```
There seems to be an error here, when testing I kept having to hit the resend button to get the email to send and when I did, it would send 2 of them. I will test it again and update this comment if it works properly.

Gehrkej replied:
Update, It still doesn't auto send, might be due to the redirect.

Gehrkej replied:
Another update, after leaving it sitting while completing the PR review, the app crashed and I received this in my terminal 
![image](https://github.com/user-attachments/assets/63286064-606e-46fc-be47-511fc5f5f9fa)

Good news is the email sent 🤷‍♂️

solderq35 replied:
I'm not sure what would have caused these errors but let me know if it is working better as of the newest commits.

Gehrkej commented on app/(auth)/signup.js:
```diff
@@ -66,6 +70,15 @@ export default function SignUp() {
 
+      try {
+        await sendEmailVerification(auth.currentUser);
+        console.log("Verification Email Sent!");
+        showSnackBar("Verification Email Sent!");
+      } catch {
+        console.log("Error sending verification email: ", e);
+        showDialog("Error", getErrorString(e));
+      }
```
It seems like the redirect event fires off before any of these logs or snackbars can be displayed
![image](https://github.com/user-attachments/assets/4d2ffac9-3fdd-4a77-a689-c3ef0b015ced)

Gehrkej commented on app/segments/(team)/chooseTeam.js:
```diff
@@ -58,9 +63,44 @@ function ChooseTeam() {
+          unregisterAuthObserver(); // Unregister the auth observer
+        } else {
+          setVerified(false);
+          console.log("Error: Email Not Verified Yet, Try Again");
```
This log along with some others spam the the console currently, unsure if we will want that when we merge this down. Might want to consolidate them a bit. 
![log spam](https://github.com/user-attachments/assets/67f190f9-f5fb-45cf-9e23-46bfcf058fa5)

Gehrkej commented on app/segments/(team)/chooseTeam.js:
The readability of what is rendered on this page is starting to get a bit tough. It doesn't have to be in this PR but it might be worth it to create a more readable implementation down the road

solderq35 replied:
Removed some of the extra logs, should be resolved by https://github.com/Golf-Drill-Challenge-App/Golf-App/pull/288/commits/69b6322cfc0d8564b927f921b64c41ba3a58fe00

Gehrkej commented on dbOperations/hooks/useUserInfo.js:
```diff
@@ -31,7 +32,14 @@ export const useUserInfo = ({
         const data = querySnapshot.data();
-        if (!data) {
+
+        // DEBUG REMOVE BELOW LOG LINES LATER
```
Ahh looks like you already have this planned lol

FrankreedX replied:
should prob remove it in this pr

solderq35 replied:
Should be resolved by https://github.com/Golf-Drill-Challenge-App/Golf-App/pull/288/commits/69b6322cfc0d8564b927f921b64c41ba3a58fe00

Gehrkej reviewed (dismissed):
Overall, the functionality works well as I am able to verify my email and sign in. We will want to figure out the Auto send functionality before merging. Additionally, for DX, we will probably want an easy way to disable this for testing purposes as merging this in will limit the amount of accounts we will have access to.

FrankreedX commented on dbOperations/hooks/useUserInfo.js:
```diff
@@ -31,7 +32,14 @@ export const useUserInfo = ({
+        console.log("VERIFIED STATUS");
+        console.log(auth.currentUser.emailVerified);
+
+        if (!data || !auth.currentUser.emailVerified) {
```
Shouldn't this use currentUserVerified instead? I think it's a nicer way of accessing this information

solderq35 replied:
Should be resolved by https://github.com/Golf-Drill-Challenge-App/Golf-App/pull/288/commits/69b6322cfc0d8564b927f921b64c41ba3a58fe00

FrankreedX commented on app/segments/(team)/chooseTeam.js:
```diff
@@ -58,9 +63,44 @@ function ChooseTeam() {
+
+  const [refreshing, setRefreshing] = useState(false);
+
+  const onRefresh = useCallback(() => {
```
Is this for debouncing? This looks like a debouncing attempt to me, but it only delay the reload everytime the user hit refresh. If that's the case, i don't think we need it.

solderq35 replied:
I removed the setTImeout from this, which is hopefully what you meant.

FrankreedX reviewed (changes requested):
Good job on this! Especially the refresh in the background stuff. It feels really responsive! I only have minor complaints, and certainly have my work cut out to merge this with my stack of PRs.

solderq35 commented:
Had to change the `context/Auth.js` useEffect from `onAuthStateChanged` to `onIdTokenChanged` in order to get `currentUserVerified` / `setCurrentUserVerified` to work as intended. Hopefully that doesn't cause too many re renders etc.

FrankreedX reviewed (approved):
lgtm
